clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -287,6 +287,7 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
             this.sender = new Sender(client,
                     this.metadata,
                     this.accumulator,
+                    config.getInt(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION) == 1,
```

Comment:
@becketqin Fair enough. I just thought it was a bit odd to change the semantics of max-in-flight-requests for the one case only. A new configuration would work also, but I agree that having a setting which is almost the same as max-in-flight-requests would be confusing for users. I was also considering whether it made sense to have the producer _always_ await batches on leadership changes, but that would make it slower to recover from hard failures when the user doesn't care about ordering. So maybe treating max-in-flight = 1 as a special case is the simplest option.